File: python/17626.py
from sys import setrecursionlimit
from math import inf

setrecursionlimit(10**6)

number = int(input())
cache = dict()


def f(x):
    if x in cache:
        return cache[x]

    if int(x**0.5) == x**0.5:
        cache[x] = 1
        return 1

    min_ = inf
    for i in range(1, x//2 + 1):
        count = f(i) + f(x-i)
        if count < min_:
            min_ = count
    cache[x] = min_
    return min_


print(f(number))

File: rust/src/6/17626.rs
use std::io::stdin;
use std::cmp::min;

fn main() {
    let number = {
        let mut input = String::new();
        stdin().read_line(&mut input).unwrap();

        input.trim().parse::<usize>().unwrap()
    };

    let mut dp = vec![0, 1];
    for i in 2..=number {
        let mut count = u32::MAX;
        let mut j = 1;
        while j*j <= i {
            count = min(count, dp[i - j*j] + 1);
            j += 1;
        }
        dp.push(count);
    }

    println!("{:?}", dp[number]);
}
